diff --git a/CMake/FollyConfigChecks.cmake b/CMake/FollyConfigChecks.cmake
index 492b20831..ee79ae4a8 100644
--- a/CMake/FollyConfigChecks.cmake
+++ b/CMake/FollyConfigChecks.cmake
@@ -98,11 +98,12 @@ check_function_exists(malloc_usable_size FOLLY_HAVE_MALLOC_USABLE_SIZE)
 
 set(CMAKE_REQUIRED_FLAGS "${FOLLY_ORIGINAL_CMAKE_REQUIRED_FLAGS}")
 
-check_cxx_source_compiles("
+check_cxx_source_runs("
   #pragma GCC diagnostic error \"-Wattributes\"
-  extern \"C\" void (*test_ifunc(void))() { return 0; }
-  void func() __attribute__((ifunc(\"test_ifunc\")));
-  int main() { return 0; }"
+  extern \"C\" int resolved_ifunc(void) { return 0; }
+  extern \"C\" int (*test_ifunc(void))() { return resolved_ifunc; }
+  int func() __attribute__((ifunc(\"test_ifunc\")));
+  int main() { return func(); }"
   FOLLY_HAVE_IFUNC
 )
 check_cxx_source_compiles("
